### PR TITLE
[Rust] prototype vec concept exercise

### PR DESCRIPTION
**Pull request by efx**
I am changing my approach to creating concept exercises. In the past I
wrote the story and filled out the documentation to suit. That incurred
burdensome costs if the narrative didn't fit the Rust idioms in the
example.rs.
So I'll try creating the example code first for us to discuss. Based on
that progress I can fill out the exercism metadata.

I took the idea from https://github.com/exercism/v3/blob/5cb717f5b77f14c64cb12fae31420830101b84b8/reference/stories/lists.language-list.md.
(#1108)

### CODE REVIEW

**Comment by coriolinus** [comment]
> I am changing my approach to creating concept exercises... I'll try creating the example code first for us to discuss.

Strongly approve of this approach. I expect it to be much more productive.

**Comment by coriolinus** [comment]
I understand this is a draft. Here are the tasks I want to see for vectors, to demonstrate understanding:

- use the `vec!` macro
- create a vector via `collect()`
- push items to the end of a vector
- pop items from the end of a vector
- index items in a vector

I think those are the minimal tasks which, when demonstrated, show that a student knows enough about vectors that they can do the rest of their research on their own.

Thinking about listy concepts, what do you think of this topology? Where "basics" is a small set including "functions", "test cases", "primitives", etc.

- Vectors requrie basics
  - sub-concepts as above
- Arrays require basics
  - array literals
  - `[T; N]` initialization
  - _cannot_ grow
- Slices require Vectors + Arrays
  - deref coersion from arrays and vectors
  - some methods?

**Comment by efx** [comment]
> Thinking about listy concepts, what do you think of this topology? Where "basics" is a small set including "functions", "test cases", "primitives", etc.

I like that topology. I find introducing the student to slices after vectors and arrays a helpful focus. Slices are important to grasp and should be handled independently.

For fun I visualized this using [dagitty](http://dagitty.net/dags.html#). As an aside, Do you know of any other way to programmatically generate DAGs using some kind of notation?
Here is how I read that topology progressing:
dagitty-model

The vectors than array connection could equally be reversed.

**Comment by coriolinus** [comment]
re: visualization: when I need graphs, I go for [graphviz](https://graphviz.org/). 

re: topology: mine differs from yours, though: in my concept, arrays do not require vectors, but slices do. It's not clear to me whether you were proposing a replacement topology, or whether that was a graphing error.

**Comment by efx** [comment]
> re: topology: mine differs from yours, though: in my concept, arrays do not require vectors, but slices do. It's not clear to me whether you were proposing a replacement topology, or whether that was a graphing error.

Ah, I believe we agree. I modeled one variation of arrays or vectors before slices. It seems like covering arrays or vectors before slices is the main thing. Not a problem though as I need to move on to working on the vectors exercise!

**Comment by gilescope** [comment]
Using extend to add two vecs might be helpful - or knowing about vec![0; 10].
I think it's helpful to see the empty initial code that the student would be presented with - looking at example.rs I'm guessing we give them the fn languages() but I'm guessing.

File: languages/rust/exercises/concept/vec/.meta/example.rs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    ]
+}
+/// return a list of languages labeled by their order
+pub fn languages_ranked() -> Vec<String> {
+    languages()
+        .iter()
+        .enumerate()
+        .map(|(index, lang)| format!("{}: {}", index + 1, lang))
+        .collect()
+}
```
Comment (coriolinus) [inline languages/rust/exercises/concept/vec/.meta/example.rs:18]: This function requires the `iterators` concept, which should probably come after `vectors`.

File: languages/rust/exercises/concept/vec/.meta/example.rs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+/// Gather your top languages to learn
+pub fn grab(which: usize) -> String {
+    if which == languages().len() {
+        languages().pop().unwrap()
+    } else {
+        languages()[which].to_string()
+    }
```
Comment (coriolinus) [inline languages/rust/exercises/concept/vec/.meta/example.rs:25]: This implementation panics if `which > languages.len()`.

File: languages/rust/exercises/concept/vec/.meta/example.rs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+}
+/// Gather your top languages to learn
+pub fn humans_and_computers(human: Vec<String>) -> Vec<String> {
+    let mut list = languages();
+    human.iter().for_each(|lang| list.push(lang.to_string()));
```
Comment (coriolinus) [inline languages/rust/exercises/concept/vec/.meta/example.rs:30]: Requires `iterators` concept.

**Comment by efx** [comment]
@gilescope 
> I think it's helpful to see the empty initial code that the student would be presented with - looking at example.rs I'm guessing we give them the fn languages() but I'm guessing.

Exactly! My understanding for the V3 exercises is that we stub function signatures with something like:
```rust

/// Gather your top languages to learn
pub fn grab(which: usize) -> String {
    unimplemented!("Solve the problem :)")
}

```
The `example.rs` represents a canonical solution used by the automated checking.

File: languages/rust/exercises/concept/vec/.meta/example.rs (hunk 1 above)
Comment (efx) [reply languages/rust/exercises/concept/vec/.meta/example.rs:18]: Great points! I'm going to rewrite the solutions using for loops. And maybe tweak the problem to not require the iterator solutions.

File: languages/rust/exercises/concept/vec/.meta/example.rs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    ]
+}
+/// Gather your top languages to learn
+pub fn grab(which: usize) -> Option<String> {
+    if which == languages().len() {
+        languages().pop()
+    } else {
+        languages().get(which).cloned()
+    }
+}
```
Comment (coriolinus) [inline languages/rust/exercises/concept/vec/.meta/example.rs:18]: This function has kind of weird semantics. As written, we could assert

```rust
assert_eq!(grab(4), Some("Scheme".into());
assert_eq!(grab(5), Some("Scheme".into());
assert_eq!(grab(6), None);
```

Comment (efx) [reply languages/rust/exercises/concept/vec/.meta/example.rs]: Removed pop semantics as I had a better idea on how to introduce that.

**Comment by efx** [comment]
@coriolinus Per learning goals, I'm going to omit "create a vector via collect()". That requires `iter` too. 
Alternatively, we could introduce `iter` much earlier.

**Comment by coriolinus** [comment]
> @coriolinus Per learning goals, I'm going to omit "create a vector via collect()". That requires `iter` too.

Yeah, good point, agree.

**Comment by efx** [comment]
Added two other simpler push and pop functions. I mocked up that and the suggested assertions here:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=1e72c9c167350473f8ac085500730a62

**Comment by coriolinus** [comment]
A bunch of comments in random order as I think about the current state of this exercise.

---

Tinkered with it a bit to mess with comments, names, implementations. Lots of bikeshedding: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=88c1ee230feae9ab55103e92a8499999

---

Going back to the initial list of requirements:

- [x] **use the vec! macro**: `fn add_one` / `fn create_single`
- [ ] ~~**create a vector via `collect()`**~~: requires `iter`
- [ ] **push items to the end of a vector**: technically covered by `fn humans_and_computers` / `fn concatenate`, but I'd communicated ambiguously. I think that `Vec::push` is more important to teach than `Vec::extend`.
- [x] **pop items from the end of a vector**: covered by `fn remove_last_one` / `fn remove_final`
- [ ] **index items in a vector**: covered by `fn grab` / `fn item`, but I'm not really happy with it. More on that later.

Overall, I'd say that this represents good progress toward the intended goals. It was my ambiguity which caused confusion about what to teach with respect to pushing. My intent is to demonstrate the use of `Vec` as an `O(1)` stack by means of `push` / `pop`. Should be pretty simple to come up with some contrived reason to use a stack.

---

The use of strings is introducing a lot of chaff with regards to cloning and borrowing. This whole exercise would be easier to implement if done on some `Copy` type such as a number. That said, it makes coming up with a story somewhat harder. I'm not committed to changing the nature of the vectors in play here.

---

I'm concerned that `fn grab` is not really suitable for the goal of teaching indexing. Your example doesn't use indexing, and a more experienced programmer might implement it as `languages().get(which).map(ToString::to_string)`, which also doesn't use indexing. It might be good to frame the index-teaching function in terms of something which requires random access to the vector's items. 

---

Here's the framing story I'd write to teach indexing: you have a `Vec<u8>` which is the contents of a binary encoded file. The encoding specification of the file is: the header consists of a list of pairs `(offset, length)`, terminated by a pair of null bytes. All data past the header is a set of arbitrary length fields. Write these functions:

- `fn num_fields(data: &Vec<u8>) -> usize`. Teaches looping through the items of a vector.
- `fn get_field(data: &Vec<u8>, field_idx: usize) -> Option<Vec<u8>>`. Teaches indexing, ranges, allocating a new vector or copying a new vector.

Pros: teaches indexing, looping, `chunks`, ranges, creating a new vector from data, can encode silly strings as binary data (i.e. "Congrats, you looked deeper 🔍👁️ ")
Cons: probably takes more than 5 minutes for a student to implement.

I'm not wedded to this idea, but I think it presents a more compelling reason to use vector methods and indices than what currently exists, and the challenge would give people a better sense of having mastered those skills.

**Comment by coriolinus** [comment]
Played around with the data-coding story; you can see it in the (not-intended-to-be-merged) branch [here](https://github.com/exercism/v3/blob/vec-indexing/languages/rust/exercises/concept/vec/src/lib.rs). I think `fn encode` is a bit much to ask of a student from a concept exercise, but `fn decode` seems simple enough that it's probably not going to ask too much from a beginner. We'd just use `encode` ourselves, to build the test cases.

They can't successfully implement `fn decode` without understanding indexing, which is the point. Student implementations might get more complex than this, because they might not know about helpers like `chunks`, but it's still only 10 lines of code; it can't be all that complicated.

**Comment by efx** [comment]
thanks for the feedback @coriolinus! 
| My intent is to demonstrate the use of Vec as an O(1) stack by means of push / pop. Should be pretty simple to come up with some contrived reason to use a stack.

I understand what you mean. I'll explore another story to capture this idea.

| The use of strings is introducing a lot of chaff with regards to cloning and borrowing. 

Yeah, it does feel like added teaching complexity. I love the idea to use integer types. That gave me an idea for a coin sleeve robot. I'll play around with that.

The `get` usage does sidestep indexing. With a numeric focus I'll try introducing basic indexing. Indexing with ranges seems like a good candidate for vec-advanced or a practice exercise.

| Played around with the data-coding story; you can see it in the (not-intended-to-be-merged) branch here. I think fn encode is a bit much to ask of a student from a concept exercise, but fn decode seems simple enough that it's probably not going to ask too much from a beginner. We'd just use encode ourselves, to build the test cases.

This exercise sounds like an excellent practice one. It requires flexing muscles acquired from concept exercises.
I feel like we can be less strict about concept overloading for practices exercises. For concept exercises it seems we should be razor focused.

**Comment by coriolinus** [comment]
> For concept exercises it seems we should be razor focused.

Proposal: to keep things simple, let's get focused. Split the `vec` concept into these concepts:

- `vec-macro` for creating empty vectors, and vectors of particular size, with`vec!`
- `vec-collect` for creating vectors via `collect()` (requires `iter`)
- `vec-stack` for using vectors as an `O(1)` stack (push and pop methods)
- `vec-index` for `O(1)` random access (incl. mutable) within a vector

To avoid overloading ourselves, let's try for these bounds:

- Not more than 15 lines of example implementation
- Not more than 3 paragraphs of documentation
- Not more than 5 tests

**Comment by gilescope** [comment]
A vec-stack exercise that did reverse polish notation calculations of numbers and operators would be neat - and also could be a nice simple example of using an enum maybe?

**Comment by efx** [comment]
Closing in favor of the decomposed approach in #3061 and following PRs.

**Pull request closed without merging**